index/inverted_index_impl.go

```diff
@@ -32,14 +34,23 @@ func NewInverted(dts ds.Batching) *invertedIndexImpl {
 }
 
 func (d *invertedIndexImpl) AddMultihashesForShard(ctx context.Context, mhIter MultihashIterator, s shard.Key) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	mhmap := make(map[string]struct{})
+
 	batch, err := d.ds.Batch(ctx)
 	if err != nil {
 		return fmt.Errorf("failed to create ds batch: %w", err)
 	}
 
 	if err := mhIter.ForEach(func(mh multihash.Multihash) error {
-		key := ds.NewKey(string(mh))
+		// if we already added a mapping for the given (multihash -> shard key) for this iterator. nothing to do here
+		if _, ok := mhmap[string(mh)]; ok {
+			return nil
+		}
```

Comment:
If you use a lock, I'm not sure that you actually need the dedup map

Reply:
We do, because a `batch,Put` wont be visible to a `ds.Get`

Reply:
Have addressed this.